src/Rings.js

```diff
@@ -0,0 +1,728 @@
+        .rangeRound([primaryMax, secondaryMax]);
+    }
+
+    secondaries.forEach(s => {
+      s.ring = 2;
+      const val = this._size ? s.size : 2;
+      s.r = this._sizeMin ? max([this._sizeMin, radiusFn(val)]) : this._sizeMax ? min([this._sizeMax, radiusFn(val)]) : radiusFn(val);
+    });
+
+    primaries.forEach(p => {
+      p.ring = 1;
+      const val = this._size ? p.size : 1;
+      p.r = this._sizeMin ? max([this._sizeMin, radiusFn(val)]) : this._sizeMax ? min([this._sizeMax, radiusFn(val)]) : radiusFn(val);
+    });
+
+    nodes = [center].concat(primaries).concat(secondaries);
+
+    primaries.forEach(p => {
+      const check = ["source", "target"];
+      const {edge} = p;
+
+      check.forEach(node => {
+        edge[node] = nodes.find(n => n.id === edge[node].id);
+      });
+
+      edges.push(edge);
+
+      linkMap[p.id].forEach(edge => {
+        const node = edge.source.id === p.id ? edge.target : edge.source;
+
+        if (node.id !== center.id) {
+          let target = secondaries.find(s => s.id === node.id);
+
+          if (!target) {
+            target = primaries.find(s => s.id === node.id);
+          }
+
+          if (target) {
+            edge.spline = true;
+
+            const centerX = width / 2;
+            const centerY = height / 2;
+            const middleRing = (secondaryRing + primaryRing) / 1.75;
+
+            const check = ["source", "target"];
+
+            check.forEach((node, i) => {
+              edge[`${node}X`] = edge[node].x + Math.cos(edge[node].ring === 2 ? edge[node].radians + Math.PI : edge[node].radians) * edge[node].r;
+              edge[`${node}Y`] = edge[node].y + Math.sin(edge[node].ring === 2 ? edge[node].radians + Math.PI : edge[node].radians) * edge[node].r;
+              edge[`${node}BisectX`] = centerX + middleRing * Math.cos(edge[node].radians);
+              edge[`${node}BisectY`] = centerY + middleRing * Math.sin(edge[node].radians);
+
+              edge[node] = nodes.find(n => n.id === edge[node].id);
+
+              if (edge[node].edges === undefined) edge[node].edges = {};
+
+              const oppId = i === 0 ? edge.target.id : edge.source.id;
+
+              if (edge[node].id === p.id) {
+                edge[node].edges[oppId] = {
+                  angle: p.radians + Math.PI,
+                  radius: ringWidth / 2
+                };
+              }
+              else {
+                edge[node].edges[oppId] = {
+                  angle: target.radians,
+                  radius: ringWidth / 2
+                };
+              }
+            });
+
+            edges.push(edge);
+          }
+        }
+      });
+    });
+
+    nodes.forEach(node => {
+
+      if (node.id !== this._center) {
+        node.rotate = node.radians * (180 / Math.PI);
+
+        const fontSize = this._shapeConfig.labelConfig.fontSize && this._shapeConfig.labelConfig.fontSize(node) || 11;
+
+        const wrap = textWrap()
+          .fontSize(fontSize);
```

Comment:
do we actually need to measure the texts here? can't we just infer the width and height from the available space from the ring sizes and that `total` variable? If we remove the text-wrapping, it might speed things up :-)